// @rocketseat/rocket-scripts/jest/index.js
/* eslint-disable global-require, import/no-dynamic-require */
const merge = require('lodash/merge');

const { existFile, realPath } = require('../utils');

const setupTestsPath = 'src/setupTests.js';
const jestConfigPath = 'jest.config.js';

const setupTests = existFile(setupTestsPath) ? [realPath(setupTestsPath)] : [];
const jestConfig = existFile(jestConfigPath) ? require(realPath(jestConfigPath)) : {};

const defaultConfig = {
  moduleFileExtensions: ['ts', 'tsx', 'js', 'jsx'],
  bail: true,
  clearMocks: true,
  coverageDirectory: 'coverage',
  coverageReporters: ['text', 'clover'],
  collectCoverageFrom: ['src/**/*.{js,jsx,ts,tsx}', '!src/**/*.d.ts'],
  coverageThreshold: {
    global: {
      branches: 80,
      functions: 80,
      lines: 80,
      statements: 80,
    },
  },
  setupFilesAfterEnv: setupTests,
  testMatch: ['**/__tests__/**/?(*.)+(spec|test).[tj]s?(x)'],
  transform: {
    '.tsx?$': 'ts-jest',
    '.jsx?$': '@sucrase/jest-plugin',
  },
  moduleNameMapper: {
    '~(.*)$': '<rootDir>/src/$1',
    '\\.(jpg|jpeg|png|gif|svg)$': '<rootDir>/__tests__/__mocks__/fileMock.js',
  },
  transformIgnorePatterns: [
    '/node_modules/+\\.js$',
  ],
};

module.exports = merge(defaultConfig, jestConfig);
