// @rocketseat/rocket-scripts/webpack/typescript/webpack.prod.js
const ClearWebpackPlugin = require('clean-webpack-plugin');
const HtmlWebpackPlugin = require('html-webpack-plugin');
const ProgressBarPlugin = require('progress-bar-webpack-plugin');
const ForkTsCheckerWebpackPlugin = require('fork-ts-checker-webpack-plugin');
const TsconfigPathsPlugin = require('tsconfig-paths-webpack-plugin');
const merge = require('webpack-merge');

const commonMainConfig = require('../webpack.common');

const { realPath } = require('../../utils');

module.exports = merge.smart(commonMainConfig, {
  entry: realPath('src/index.tsx'),
  mode: 'production',
  target: 'web',
  output: {
    filename: 'bundle.[name].js',
    chunkFilename: 'chunk.[name].js',
  },
  module: {
    rules: [
      {
        test: /\.tsx?$/,
        loader: 'ts-loader',
        options: {
          transpileOnly: true,
        },
      },
    ],
  },
  plugins: [
    new ClearWebpackPlugin(),
    new HtmlWebpackPlugin({
      template: realPath('public/index.html'),
    }),
    new ForkTsCheckerWebpackPlugin({
      tsconfig: realPath('tsconfig.json'),
      useTypescriptIncrementalApi: true,
      checkSyntacticErrors: true,
      reportFiles: ['**', '!**/__tests__/**', '!**/?(*.)(spec|test).*'],
    }),
    new ProgressBarPlugin({
      summary: false,
    }),
  ],
  resolve: {
    extensions: ['.ts', '.tsx'],
    plugins: [new TsconfigPathsPlugin()],
  },
  devtool: 'source-map',
});
